Rando: Master Sword Shuffle

stratomaster64 opened the pull request:
Adds the Master Sword to the item pool, requiring Adult Link to be more creative in his offensive strategies until he finds it.
Because of this:

- Sheik/Ganondorf will give you hints on where the Master Sword may be located.
- Logic is updated to reflect new requirements.
- If you manage to get to Ganon's boss fight without MS, the fight will not give it to you (which fixes a vanilla bug).
- Swordless Link oddities have also been addressed.

<!--- section:artifacts:start -->
### Build Artifacts
  - [soh.otr.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/988304183.zip)
  - [soh-linux-compatibility.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/988304184.zip)
  - [soh-linux-performance.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/988304185.zip)
  - [soh-mac.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/988304186.zip)
  - [soh-switch.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/988304187.zip)
  - [soh-wiiu.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/988304189.zip)
  - [soh-windows.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/988304190.zip)
<!--- section:artifacts:end -->

stratomaster64 committed:
The mother of all commits

stratomaster64 commented:
Known issues so far:

- [x] Spoiler log doesn't see Master Sword as an item required to beat the game (for Ganon) and still assume you obtain the Master Sword by turning adult.
- [x] Two "Master Sword" entries are placed into the world, expected only one.
  - [x] Actually obtaining the Master Sword at these locations will present itself as a different item entirely (due to the `GetItem` enum for Master Sword). Removing the errant enum results in the "Master Sword" looking like a BGS (expected), but not actually giving the sword itself, instead giving bombs on the item slot.
- [x] Although starting as Adult Link correctly leaves you swordless until you obtain a sword, starting as child and turning adult still gives you the Master Sword.

stratomaster64 committed:
Removed `GI_SWORD_MASTER`;
"Master Sword" Items now actually give MS

stratomaster64 committed:
Removed dupe MS entries in item pool;
updated GIMESSAGE (should stop crashing on non-Windows);
re-added MS in item list

stratomaster64 committed:
Give Adult Link a freebie with shuffle MS on;
cihld -> adult no longer gives MS;
ToT Master Sword now gives correct item

Archez committed:
add master sword GI draw func based on ToT MS object

stratomaster64 committed:
Merge pull request #8 from Archez/master-sword-gi-draw

Add Master Sword GI draw func based on ToT MS object

stratomaster64 committed:
Force `MasterSword` logic var to only update upon getting MS

stratomaster64 committed:
Dorf funny line now activates with LA and MS in inv

Archez commented on soh/soh/Enhancements/randomizer/3drando/logic.cpp:
```diff
@@ -512,7 +514,7 @@ namespace Logic {
     GoldScale       = ProgressiveScale      >= 2;
     AdultsWallet    = ProgressiveWallet     >= 1;
-    MasterSword     = MasterSword   || IsAdult;
+    MasterSword     = MasterSword;
```
This does nothing and can just be removed entirely.

`MasterSword` is updated appropriately via the item table logic, similar to all other item logic variables.

Archez commented on soh/soh/Enhancements/randomizer/3drando/item_pool.cpp:
```diff
@@ -47,7 +47,7 @@ const std::array<uint32_t, 16> JunkPoolItems = {
   ICE_TRAP,
 };
-const std::array<uint32_t, 59> alwaysItems = {
+const std::array<uint32_t, 60> alwaysItems = {
```
Nothing was added to this vector, please set the size back.

Archez commented on soh/soh/Enhancements/randomizer/3drando/item_pool.cpp:
```diff
@@ -108,7 +108,7 @@ const std::array<uint32_t, 59> alwaysItems = {
   TREASURE_GAME_HEART,
 };
-const std::array<uint32_t, 43> easyItems = {
+const std::array<uint32_t, 44> easyItems = {
```
You previously had the master sword added to this array, but then removed it. Please add it back. This is the "plentiful" item pool, and so this would make it so there are 2 possible master swords in the item pool for plentiful (matching what 3ds does).

Archez commented on soh/src/code/z_parameter.c:
```diff
@@ -1415,13 +1415,18 @@ void Inventory_SwapAgeEquipment(void) {
         // When becoming adult, remove swordless flag since we'll get master sword
         // Only in rando to keep swordless link bugs in vanilla
         if (gSaveContext.n64ddFlag) {
+            if (!Randomizer_GetSettingValue(RSK_SHUFFLE_MASTER_SWORD)) {
```
Maybe combine these if statements together and update the comment above regarding MS shuffle

Archez commented on soh/src/code/z_parameter.c:
```diff
@@ -1415,13 +1415,18 @@ void Inventory_SwapAgeEquipment(void) {
 
         if (gSaveContext.adultEquips.buttonItems[0] == ITEM_NONE) {
-            gSaveContext.equips.buttonItems[0] = ITEM_SWORD_MASTER;
+            if (gSaveContext.n64ddFlag && Randomizer_GetSettingValue(RSK_SHUFFLE_MASTER_SWORD)) {;}
+            else {
+                gSaveContext.equips.buttonItems[0] = ITEM_SWORD_MASTER;
+            }
```
The empty block and else condition feels wrong. You could just invert the conditional to avoid the else case.

```suggestion
            if (!gSaveContext.n64ddFlag || !Randomizer_GetSettingValue(RSK_SHUFFLE_MASTER_SWORD)) {
                gSaveContext.equips.buttonItems[0] = ITEM_SWORD_MASTER;
            }
```

Archez commented on soh/src/code/z_sram.c:
```diff
@@ -168,10 +168,13 @@ void Sram_OpenSave() {
-        gSaveContext.equips.buttonItems[0] = ITEM_SWORD_MASTER;
-        gSaveContext.equips.equipment &= ~0xF;
-        gSaveContext.equips.equipment |= 2;
+        if (gSaveContext.n64ddFlag && Randomizer_GetSettingValue(RSK_SHUFFLE_MASTER_SWORD)) {;}
+        else {
+            gSaveContext.inventory.equipment |= gBitFlags[1] << gEquipShifts[EQUIP_SWORD];
+            gSaveContext.equips.buttonItems[0] = ITEM_SWORD_MASTER;
+            gSaveContext.equips.equipment &= ~0xF;
+            gSaveContext.equips.equipment |= 2;
+        }
```
Same thing here.

```suggestion
        if (!gSaveContext.n64ddFlag || !Randomizer_GetSettingValue(RSK_SHUFFLE_MASTER_SWORD)) {
            gSaveContext.inventory.equipment |= gBitFlags[1] << gEquipShifts[EQUIP_SWORD];
            gSaveContext.equips.buttonItems[0] = ITEM_SWORD_MASTER;
            gSaveContext.equips.equipment &= ~0xF;
            gSaveContext.equips.equipment |= 2;
        }
```

Archez reviewed (changes requested):
Looking good so far. I've skipped over the `locacc_xxx` logic for now, hoping others can chime in there.

Something else that stands out, you most likely need to make changes in `z_boss_ganon2` to prevent the player from getting a free MS sword. Could probably "kill" the spawned sword, or something.

stratomaster64 committed:
Apply suggestions

inspectredc commented on soh/src/code/z_parameter.c:
```diff
@@ -1413,15 +1413,18 @@ void Inventory_SwapAgeEquipment(void) {
+        // (Unless Master Sword is shuffled)
         // Only in rando to keep swordless link bugs in vanilla
-        if (gSaveContext.n64ddFlag) {
+        if (!gSaveContext.n64ddFlag || !Randomizer_GetSettingValue(RSK_SHUFFLE_MASTER_SWORD)) {
```
Not sure I understand why this is being changed liked this, surely we want

`if (gSaveContext.n64ddFlag && !Randomizer_GetSettingValue(RSK_SHUFFLE_MASTER_SWORD))`

stratomaster64 replied:
What I wrote is along the lines of `if the file is not randomizer, or the randomizer setting for shuffling MS is OFF` (which can only be set in a rando file)
I wasn't too sure about how to write it in a positive case, but your suggestion would fail in vanilla since it's written as `if the file is randomizer and the randomizer setting for shuffling MS is OFF`. Unless having MS on the B button automatically clears the swordless flag?

inspectredc replied:
In vanilla I believe the aim is to retain the bugs it has regarding change age. The swordless flag is never changed when turning adult normally I believe, it is only removed when selecting a sword in the equipment menu

Archez replied:
Yeah, the original suggestion I had for "combining the if statements" was to look like what @inspectredc  posted above.

The other two sets of if statements were fine because we wanted to inverse them to avoid the `{;}`

stratomaster64 committed:
Updated RAND_INF;
Check Tracker changes;
Gave RAND_INF and ice trap logic to ToT MS check;
Fixed swordless behavior for HBA/fishing

inspectredc commented:
We might want something along the lines of 

```
if (gSaveContext.n64ddFlag && Randomizer_GetSettingValue(RSK_SHUFFLE_MASTER_SWORD) && gSaveContext.equips.buttonItems[0] == ITEM_NONE) {
    gSaveContext.equips.equipment = 0x1120;
} else {
    gSaveContext.equips.equipment = 0x1122;
}
```

where this section is https://github.com/HarbourMasters/Shipwright/blob/778bf5a0e36b97ec520cbe4795041ab0b8c762eb/soh/src/code/z_parameter.c#L1440

I think what this does is stops the master sword slot from being highlighted in the pause menu but please double check

Commit:
Merge branch 'HarbourMasters:develop' into mister-sword

stratomaster64 commented:
> We might want something along the lines of
> 
> ```
> if (gSaveContext.n64ddFlag && Randomizer_GetSettingValue(RSK_SHUFFLE_MASTER_SWORD) && gSaveContext.equips.buttonItems[0] == ITEM_NONE) {
>     gSaveContext.equips.equipment = 0x1120;
> } else {
>     gSaveContext.equips.equipment = 0x1122;
> }
> ```
> 
> where this section is
> 
> https://github.com/HarbourMasters/Shipwright/blob/778bf5a0e36b97ec520cbe4795041ab0b8c762eb/soh/src/code/z_parameter.c#L1440
> 
> I think what this does is stops the master sword slot from being highlighted in the pause menu but please double check

Works as expected, will include in the next commit. Thanks!

stratomaster64 committed:
ToT MS Check now works in check tracker;
Visual bug where box hovers over non-existent MS gone;
Fixed RAND_INF check with ToT MS pedestal;
Ganon no longer gives free MS

inspectredc commented on soh/src/code/z_parameter.c:
```diff
@@ -1415,15 +1437,18 @@ void Inventory_SwapAgeEquipment(void) {
 
         if (gSaveContext.adultEquips.buttonItems[0] == ITEM_NONE) {
-            gSaveContext.equips.buttonItems[0] = ITEM_SWORD_MASTER;
+            if (!gSaveContext.n64ddFlag || !Randomizer_GetSettingValue(RSK_SHUFFLE_MASTER_SWORD)) {
+                gSaveContext.equips.buttonItems[0] = ITEM_SWORD_MASTER;
+            }
```
```suggestion
            if (!gSaveContext.n64ddFlag || !Randomizer_GetSettingValue(RSK_SHUFFLE_MASTER_SWORD)) {
                gSaveContext.equips.buttonItems[0] = ITEM_SWORD_MASTER;
            } else {
                gSaveContext.equips.buttonItems[0] = ITEM_NONE;
            }
```

I believe this is needed otherwise if child link has kokiri sword equipped then adult link will have kokiri sword

inspectredc commented on soh/src/code/z_parameter.c:
```diff
@@ -1415,15 +1437,18 @@ void Inventory_SwapAgeEquipment(void) {
 
         gSaveContext.childEquips.equipment = gSaveContext.equips.equipment;
 
         if (gSaveContext.adultEquips.buttonItems[0] == ITEM_NONE) {
```
The current reason for adult c-button items not saving when without the master sword is because it re-enters this condition https://github.com/HarbourMasters/Shipwright/blob/b848eb9b83532e3d47fed69184cd48718b906fac/soh/src/code/z_parameter.c#L1448 May want to expand this so that it checks that adult equips is initialised.

e.g. 
```
if (gSaveContext.adultEquips.buttonItems[0] == ITEM_NONE && 
    !(gSaveContext.n64ddFlag && Randomizer_GetSettingValue(RSK_SHUFFLE_MASTER_SWORD) && gSaveContext.adultEquips.equipment != 0))
```

though there is probably a better way for checking if you haven't been adult yet and this looks kinda messy

stratomaster64 committed:
adult equips no longer reset in MS shuffle

Commit:
Merge branch 'HarbourMasters:develop' into mister-sword

stratomaster64 commented:
@inspectredc Could you also look into the location access logic files to confirm that everything makes sense there? I mostly copied off 3d rando changes for their MS shuffle, except where you already modified logic for tricks.

inspectredc commented:
> @inspectredc Could you also look into the location access logic files to confirm that everything makes sense there? I mostly copied off 3d rando changes for their MS shuffle, except where you already modified logic for tricks.

Yeah will do that today 👍🏼

inspectredc commented on soh/soh/Enhancements/randomizer/3drando/location_access/locacc_fire_temple.cpp:
```diff
@@ -64,7 +64,8 @@ void AreaTable_Init_FireTemple() {
                 }, {
```
https://github.com/HarbourMasters/Shipwright/blob/af1761da109baf72fdb57433622dfe261642402a/soh/soh/Enhancements/randomizer/3drando/location_access/locacc_fire_temple.cpp#L63

To be consistent with how the logic is for defeating a flare dancer later, this should be something like 

```
LocationAccess(FIRE_TEMPLE_FLARE_DANCER_CHEST, {[]{return IsAdult && (CanUse(MEGATON_HAMMER) || CanUse(HOOKSHOT) || (HasExplosives && (CanUse(MASTER_SWORD) || CanUse(KOKIRI_SWORD) || CanUse(BIGGORON_SWORD) || CanUse(SLINGSHOT) || CanUse(BOW) || CanUse(BOOMERANG))));}}),
```

The IsAdult is still needed to actually reach the chest

inspectredc commented on soh/soh/Enhancements/randomizer/3drando/location_access/locacc_fire_temple.cpp:
```diff
@@ -64,7 +64,8 @@ void AreaTable_Init_FireTemple() {
                   Entrance(FIRE_TEMPLE_LOOP_TILES,         {[]{return true;}}),
-                  Entrance(FIRE_TEMPLE_LOOP_HAMMER_SWITCH, {[]{return Here(FIRE_TEMPLE_LOOP_FLARE_DANCER, []{return (HasExplosives || CanUse(MEGATON_HAMMER) || CanUse(HOOKSHOT)) && (IsAdult || KokiriSword || Slingshot || Boomerang);});}}),
+                  Entrance(FIRE_TEMPLE_LOOP_HAMMER_SWITCH, {[]{return Here(FIRE_TEMPLE_LOOP_FLARE_DANCER, []{return (HasExplosives || CanUse(MEGATON_HAMMER) || CanUse(HOOKSHOT)) && (CanUse(KOKIRI_SWORD) || CanUse(MASTER_SWORD) || CanUse(BIGGORON_SWORD) ||
+                   CanUse(SLINGSHOT) || CanUse(BOOMERANG));});}}),
```
```suggestion
                  Entrance(FIRE_TEMPLE_LOOP_HAMMER_SWITCH, {[]{return Here(FIRE_TEMPLE_LOOP_FLARE_DANCER, []{return CanUse(MEGATON_HAMMER) || CanUse(HOOKSHOT) || (HasExplosives && (CanUse(KOKIRI_SWORD) || CanUse(MASTER_SWORD) || CanUse(BIGGORON_SWORD) ||
                   CanUse(SLINGSHOT) || CanUse(BOW) || CanUse(BOOMERANG)));});}}),
```
I've put the explanation for this with the upper flare dancer

inspectredc commented on soh/soh/Enhancements/randomizer/3drando/location_access/locacc_fire_temple.cpp:
```diff
@@ -276,8 +277,10 @@ void AreaTable_Init_FireTemple() {
+                  Entrance(FIRE_TEMPLE_LATE_FIRE_MAZE, {[]{return Here(FIRE_TEMPLE_UPPER_FLARE_DANCER, []{return (HasExplosives || CanUse(MEGATON_HAMMER) || CanUse(HOOKSHOT)) && ((CanUse(KOKIRI_SWORD)) || CanUse(MASTER_SWORD) || CanUse(BIGGORON_SWORD) ||
+                   CanUse(SLINGSHOT) || CanUse(BOOMERANG));});}}),
+                  Entrance(FIRE_TEMPLE_WEST_CLIMB,     {[]{return Here(FIRE_TEMPLE_UPPER_FLARE_DANCER, []{return (HasExplosives || CanUse(MEGATON_HAMMER) || CanUse(HOOKSHOT)) && ((CanUse(KOKIRI_SWORD)) || CanUse(MASTER_SWORD) || CanUse(BIGGORON_SWORD) ||
+                   CanUse(SLINGSHOT) || CanUse(BOOMERANG));});}}),
```
```suggestion
                  Entrance(FIRE_TEMPLE_LATE_FIRE_MAZE, {[]{return Here(FIRE_TEMPLE_UPPER_FLARE_DANCER, []{return CanUse(MEGATON_HAMMER) || CanUse(HOOKSHOT) || (HasExplosives && ((CanUse(KOKIRI_SWORD)) || CanUse(MASTER_SWORD) || CanUse(BIGGORON_SWORD) ||
                   CanUse(SLINGSHOT) || CanUse(BOW) || CanUse(BOOMERANG)));});}}),
                  Entrance(FIRE_TEMPLE_WEST_CLIMB,     {[]{return Here(FIRE_TEMPLE_UPPER_FLARE_DANCER, []{return CanUse(MEGATON_HAMMER) || CanUse(HOOKSHOT) || (HasExplosives && ((CanUse(KOKIRI_SWORD)) || CanUse(MASTER_SWORD) || CanUse(BIGGORON_SWORD) ||
                   CanUse(SLINGSHOT) || CanUse(BOOMERANG)));});}}),
```

Flare dancers can be beaten just with hammer or just with hookshot (technically also with just explosives but theyre consumable)

inspectredc commented on soh/soh/Enhancements/randomizer/3drando/location_access/locacc_fire_temple.cpp:
```diff
@@ -330,7 +333,7 @@ void AreaTable_Init_FireTemple() {
   }, {
                   //Locations
-                  LocationAccess(FIRE_TEMPLE_MQ_MEGATON_HAMMER_CHEST, {[]{return IsAdult && (HasExplosives || Hammer || Hookshot);}}),
+                  LocationAccess(FIRE_TEMPLE_MQ_MEGATON_HAMMER_CHEST, {[]{return IsAdult && (CanUse(KOKIRI_SWORD) || CanUse(MASTER_SWORD) || CanUse(BIGGORON_SWORD)) && (HasExplosives || Hammer || Hookshot);}}),
```
```suggestion
                  LocationAccess(FIRE_TEMPLE_MQ_MEGATON_HAMMER_CHEST, {[]{return IsAdult && (CanUse(MEGATON_HAMMER) || CanUse(HOOKSHOT) || (HasExplosives && (CanUse(MASTER_SWORD) || CanUse(KOKIRI_SWORD) || CanUse(BIGGORON_SWORD) || CanUse(BOW) || CanUse(SLINGSHOT) || CanUse(BOOMERANG))));}}),
```

Same as flare dancers before, this is less necessary since in glitchless you still need a sword to reach here but makes sense to have the change done properly anyway

inspectredc commented on soh/soh/Enhancements/randomizer/3drando/location_access/locacc_fire_temple.cpp:
```diff
@@ -353,7 +356,7 @@ void AreaTable_Init_FireTemple() {
   areaTable[FIRE_TEMPLE_MQ_LOWER_MAZE] = Area("Fire Temple MQ Lower Maze", "Fire Temple", FIRE_TEMPLE, NO_DAY_NIGHT_CYCLE, {}, {
                   //Locations
-                  LocationAccess(FIRE_TEMPLE_MQ_LIZALFOS_MAZE_LOWER_CHEST,     {[]{return true;}}),
+                  LocationAccess(FIRE_TEMPLE_MQ_LIZALFOS_MAZE_LOWER_CHEST,     {[]{return CanUse(MASTER_SWORD) || CanUse(BIGGORON_SWORD);}}),
```
```suggestion
                  LocationAccess(FIRE_TEMPLE_MQ_LIZALFOS_MAZE_LOWER_CHEST,     {[]{return CanUse(MASTER_SWORD) || CanUse(KOKIRI_SWORD) || CanUse(BIGGORON_SWORD);}}),
```

inspectredc commented on soh/soh/Enhancements/randomizer/3drando/location_access/locacc_ganons_castle.cpp:
```diff
@@ -87,7 +87,7 @@ void AreaTable_Init_GanonsCastle() {
   areaTable[GANONS_CASTLE_SPIRIT_TRIAL] = Area("Ganon's Castle Spirit Trial", "Ganon's Castle", GANONS_CASTLE, NO_DAY_NIGHT_CYCLE, {
                   //Events
                   EventAccess(&NutPot,           {[]{return NutPot || ((LogicSpiritTrialHookshot || CanUse(HOOKSHOT)) && HasBombchus && Bow && (MirrorShield || (SunlightArrows && CanUse(LIGHT_ARROWS))) && IsAdult);}}),
```
```suggestion
                  EventAccess(&NutPot,           {[]{return NutPot || (((LogicSpiritTrialHookshot && CanJumpslash) || CanUse(HOOKSHOT)) && HasBombchus && CanUse(BOW) && (CanUse(MIRROR_SHIELD) || (SunlightArrows && CanUse(LIGHT_ARROWS))));}}),
```

inspectredc commented on soh/soh/Enhancements/randomizer/3drando/location_access/locacc_ganons_castle.cpp:
```diff
@@ -87,7 +87,7 @@ void AreaTable_Init_GanonsCastle() {
                   //Events
                   EventAccess(&NutPot,           {[]{return NutPot || ((LogicSpiritTrialHookshot || CanUse(HOOKSHOT)) && HasBombchus && Bow && (MirrorShield || (SunlightArrows && CanUse(LIGHT_ARROWS))) && IsAdult);}}),
-                  EventAccess(&SpiritTrialClear, {[]{return CanUse(LIGHT_ARROWS)  && (MirrorShield || SunlightArrows) && HasBombchus && (LogicSpiritTrialHookshot || CanUse(HOOKSHOT));}}),
+                  EventAccess(&SpiritTrialClear, {[]{return CanUse(LIGHT_ARROWS)  && (IsAdult && CanUse(MIRROR_SHIELD) || SunlightArrows) && HasBombchus && (LogicSpiritTrialHookshot || CanUse(HOOKSHOT));}}),
```
```suggestion
                  EventAccess(&SpiritTrialClear, {[]{return CanUse(LIGHT_ARROWS) && (CanUse(MIRROR_SHIELD) || SunlightArrows) && HasBombchus && ((LogicSpiritTrialHookshot && CanJumpslash) || CanUse(HOOKSHOT));}}),
```

spirit clear can be done as child, just now requires an extra check for being able to jumpslash to achieve the logic trick

inspectredc commented on soh/soh/Enhancements/randomizer/3drando/location_access/locacc_ganons_castle.cpp:
```diff
@@ -153,12 +153,12 @@ void AreaTable_Init_GanonsCastle() {
   areaTable[GANONS_CASTLE_MQ_FOREST_TRIAL] = Area("Ganon's Castle MQ Forest Trial", "Ganons Castle", GANONS_CASTLE, NO_DAY_NIGHT_CYCLE, {
                   //Events
-                  EventAccess(&ForestTrialClear, {[]{return IsAdult && CanUse(LIGHT_ARROWS) && CanPlay(SongOfTime);}}),
+                  EventAccess(&ForestTrialClear, {[]{return IsAdult && CanUse(LIGHT_ARROWS) && (CanUse(KOKIRI_SWORD) || CanUse(MASTER_SWORD) || CanUse(BIGGORON_SWORD)) &&CanPlay(SongOfTime);}}),
```
```suggestion
                  EventAccess(&ForestTrialClear, {[]{return IsAdult && CanUse(LIGHT_ARROWS) && (CanUse(KOKIRI_SWORD) || CanUse(MASTER_SWORD) || CanUse(BIGGORON_SWORD)) && CanPlay(SongOfTime);}}),
```

inspectredc commented on soh/soh/Enhancements/randomizer/3drando/location_access/locacc_ganons_castle.cpp:
```diff
@@ -203,11 +203,11 @@ void AreaTable_Init_GanonsCastle() {
   areaTable[GANONS_CASTLE_MQ_LIGHT_TRIAL] = Area("Ganon's Castle MQ Light Trial", "Ganons Castle", GANONS_CASTLE, NO_DAY_NIGHT_CYCLE, {
                   //Events
-                  EventAccess(&LightTrialClear, {[]{return IsAdult && CanUse(LIGHT_ARROWS) && SmallKeys(GANONS_CASTLE, 3) && (LogicLensCastleMQ || CanUse(LENS_OF_TRUTH)) && (Hookshot || LogicLightTrialMQ);}}),
+                  EventAccess(&LightTrialClear, {[]{return IsAdult && (CanUse(MASTER_SWORD) || CanUse(BIGGORON_SWORD)) && CanUse(LIGHT_ARROWS) && SmallKeys(GANONS_CASTLE, 3) && (LogicLensCastleMQ || CanUse(LENS_OF_TRUTH)) && (Hookshot || LogicLightTrialMQ);}}),
```
```suggestion
                  EventAccess(&LightTrialClear, {[]{return IsAdult && (CanUse(MASTER_SWORD) || CanUse(KOKIRI_SWORD) || CanUse(BIGGORON_SWORD)) && CanUse(LIGHT_ARROWS) && SmallKeys(GANONS_CASTLE, 3) && (LogicLensCastleMQ || CanUse(LENS_OF_TRUTH)) && (Hookshot || LogicLightTrialMQ);}}),
```

inspectredc commented on soh/soh/Enhancements/randomizer/3drando/location_access/locacc_ganons_castle.cpp:
```diff
@@ -203,11 +203,11 @@ void AreaTable_Init_GanonsCastle() {
   }, {
                   //Locations
-                  LocationAccess(GANONS_CASTLE_MQ_LIGHT_TRIAL_LULLABY_CHEST, {[]{return CanPlay(ZeldasLullaby);}}),
+                  LocationAccess(GANONS_CASTLE_MQ_LIGHT_TRIAL_LULLABY_CHEST, {[]{return (CanUse(MASTER_SWORD) || CanUse(BIGGORON_SWORD)) && CanPlay(ZeldasLullaby);}}),
```
```suggestion
                  LocationAccess(GANONS_CASTLE_MQ_LIGHT_TRIAL_LULLABY_CHEST, {[]{return (CanUse(MASTER_SWORD) || CanUse(KOKIRI_SWORD) || CanUse(BIGGORON_SWORD)) && CanPlay(ZeldasLullaby);}}),
```

inspectredc commented on soh/soh/Enhancements/randomizer/3drando/location_access/locacc_gerudo_training_grounds.cpp:
```diff
@@ -135,10 +135,10 @@ void AreaTable_Init_GerudoTrainingGrounds() {
   }, {
                   //Locations
-                  LocationAccess(GERUDO_TRAINING_GROUNDS_MQ_DINOLFOS_CHEST, {[]{return IsAdult;}}),
+                  LocationAccess(GERUDO_TRAINING_GROUNDS_MQ_DINOLFOS_CHEST, {[]{return IsAdult && (CanUse(MASTER_SWORD) || CanUse(BIGGORON_SWORD));}}),
```
```suggestion
                  LocationAccess(GERUDO_TRAINING_GROUNDS_MQ_DINOLFOS_CHEST, {[]{return IsAdult && (CanUse(MASTER_SWORD) || CanUse(KOKIRI_SWORD) || CanUse(BIGGORON_SWORD));}}),
```

inspectredc commented on soh/soh/Enhancements/randomizer/3drando/location_access/locacc_gerudo_training_grounds.cpp:
```diff
@@ -135,10 +135,10 @@ void AreaTable_Init_GerudoTrainingGrounds() {
   }, {
                   //Exits
-                  Entrance(GERUDO_TRAINING_GROUNDS_MQ_UNDERWATER, {[]{return (Bow || CanUse(LONGSHOT)) && CanUse(HOVER_BOOTS) && IsAdult;}}),
+                  Entrance(GERUDO_TRAINING_GROUNDS_MQ_UNDERWATER, {[]{return (Bow || (CanUse(LONGSHOT) && HasFireSource)) && CanUse(HOVER_BOOTS) && IsAdult && (CanUse(MASTER_SWORD) || CanUse(BIGGORON_SWORD));}}),
```
```suggestion
                  Entrance(GERUDO_TRAINING_GROUNDS_MQ_UNDERWATER, {[]{return (Bow || (CanUse(LONGSHOT) && HasFireSource)) && CanUse(HOVER_BOOTS) && IsAdult && (CanUse(MASTER_SWORD) || CanUse(KOKIRI_SWORD) || CanUse(BIGGORON_SWORD));}}),
```

inspectredc commented on soh/soh/Enhancements/randomizer/3drando/location_access/locacc_gerudo_training_grounds.cpp:
```diff
@@ -160,8 +160,8 @@ void AreaTable_Init_GerudoTrainingGrounds() {
-                  LocationAccess(GERUDO_TRAINING_GROUNDS_MQ_BEFORE_HEAVY_BLOCK_CHEST, {[]{return IsAdult;}}),
-                  LocationAccess(GERUDO_TRAINING_GROUNDS_MQ_HEAVY_BLOCK_CHEST,        {[]{return CanUse(SILVER_GAUNTLETS);}}),
+                  LocationAccess(GERUDO_TRAINING_GROUNDS_MQ_BEFORE_HEAVY_BLOCK_CHEST, {[]{return IsAdult && (CanUse(MASTER_SWORD) || CanUse(BIGGORON_SWORD));}}),
+                  LocationAccess(GERUDO_TRAINING_GROUNDS_MQ_HEAVY_BLOCK_CHEST,        {[]{return CanUse(SILVER_GAUNTLETS) && (CanUse(MASTER_SWORD) || CanUse(BIGGORON_SWORD));}}),
```
```suggestion
                  LocationAccess(GERUDO_TRAINING_GROUNDS_MQ_BEFORE_HEAVY_BLOCK_CHEST, {[]{return IsAdult && (CanUse(MASTER_SWORD) || CanUse(KOKIRI_SWORD) || CanUse(BIGGORON_SWORD));}}),
                  LocationAccess(GERUDO_TRAINING_GROUNDS_MQ_HEAVY_BLOCK_CHEST,        {[]{return CanUse(SILVER_GAUNTLETS) && (CanUse(MASTER_SWORD) || CanUse(KOKIRI_SWORD) || CanUse(BIGGORON_SWORD));}}),
```

inspectredc commented on soh/soh/Enhancements/randomizer/3drando/location_access/locacc_gerudo_training_grounds.cpp:
```diff
@@ -170,12 +170,12 @@ void AreaTable_Init_GerudoTrainingGrounds() {
-                  LocationAccess(GERUDO_TRAINING_GROUNDS_MQ_SECOND_IRON_KNUCKLE_CHEST, {[]{return true;}}),
-                  LocationAccess(GERUDO_TRAINING_GROUNDS_MQ_FLAME_CIRCLE_CHEST,        {[]{return CanUse(HOOKSHOT) || Bow || HasExplosives;}}),
+                  LocationAccess(GERUDO_TRAINING_GROUNDS_MQ_EYE_STATUE_CHEST,          {[]{return CanUse(BOW);}}),
+                  LocationAccess(GERUDO_TRAINING_GROUNDS_MQ_SECOND_IRON_KNUCKLE_CHEST, {[]{return (CanUse(MASTER_SWORD) || CanUse(BIGGORON_SWORD));}}),
```
```suggestion
                  LocationAccess(GERUDO_TRAINING_GROUNDS_MQ_SECOND_IRON_KNUCKLE_CHEST, {[]{return (CanUse(MASTER_SWORD) || CanUse(KOKIRI_SWORD) || CanUse(BIGGORON_SWORD));}}),
```

inspectredc commented on soh/soh/Enhancements/randomizer/3drando/location_access/locacc_gerudo_valley.cpp:
```diff
@@ -104,13 +104,13 @@ void AreaTable_Init_GerudoValley() {
                   LocationAccess(GF_GERUDO_MEMBERSHIP_CARD,       {[]{return CanFinishGerudoFortress;}}),
                   LocationAccess(GF_GS_ARCHERY_RANGE,   {[]{return IsAdult && HookshotOrBoomerang && GerudoToken && AtNight && CanGetNightTimeGS;}}),
-                  LocationAccess(GF_GS_TOP_FLOOR,       {[]{return IsAdult && AtNight && (GerudoToken || CanUse(BOW) || CanUse(HOOKSHOT) || CanUse(HOVER_BOOTS) || LogicGerudoKitchen || LogicGFJump) && CanGetNightTimeGS;}}),
+                  LocationAccess(GF_GS_TOP_FLOOR,       {[]{return IsAdult && AtNight && CanAdultAttack && (GerudoToken || CanUse(BOW) || CanUse(HOOKSHOT) || CanUse(HOVER_BOOTS) || LogicGerudoKitchen || LogicGFJump) && CanGetNightTimeGS;}}),
```
Can't use CanAdultAttack here since hammer cannot reach this skull

inspectredc replied:
probably can leave this stuff for my logic cleanup PR since it happens quite a lot even ignoring these changes

inspectredc commented on soh/soh/Enhancements/randomizer/3drando/location_access/locacc_deku_tree.cpp:
```diff
@@ -70,14 +70,14 @@ void AreaTable_Init_DekuTree() {
                   LocationAccess(DEKU_TREE_BASEMENT_CHEST,    {[]{return true;}}),
-                  LocationAccess(DEKU_TREE_GS_BASEMENT_GATE,  {[]{return IsAdult || CanChildAttack;}}),
-                  LocationAccess(DEKU_TREE_GS_BASEMENT_VINES, {[]{return CanUseProjectile || CanUse(DINS_FIRE) || (LogicDekuBasementGS && (IsAdult || Sticks || KokiriSword));}}),
+                  LocationAccess(DEKU_TREE_GS_BASEMENT_GATE,  {[]{return CanAdultAttack || CanChildAttack;}}),
```
This needs to be changed since hammer cannot reach this skull

inspectredc replied:
```suggestion
                  LocationAccess(DEKU_TREE_GS_BASEMENT_GATE,  {[]{return CanJumpslash || CanUse(SLINGSHOT) || CanUse(BOOMERANG) || HasExplosives || CanUse(BOW) || CanUse(HOOKSHOT) || CanUse(DINS_FIRE);}}),
```

Just wanted to re-highlight the issue here with not being able to get this skulltula with a hammer, and provide a suggestion

inspectredc commented on soh/soh/Enhancements/randomizer/3drando/location_access/locacc_gerudo_valley.cpp:
```diff
@@ -172,7 +172,7 @@ void AreaTable_Init_GerudoValley() {
                   LocationAccess(COLOSSUS_GS_BEAN_PATCH,    {[]{return CanPlantBugs && CanChildAttack;}}),
                   LocationAccess(COLOSSUS_GS_TREE,          {[]{return IsAdult && HookshotOrBoomerang && AtNight && CanGetNightTimeGS;}}),
-                  LocationAccess(COLOSSUS_GS_HILL,          {[]{return IsAdult && AtNight && (CanPlantBean(DESERT_COLOSSUS) || CanUse(LONGSHOT) || (LogicColossusGS && CanUse(HOOKSHOT))) && CanGetNightTimeGS;}}),
+                  LocationAccess(COLOSSUS_GS_HILL,          {[]{return IsAdult && CanAdultAttack && AtNight && (CanPlantBean(DESERT_COLOSSUS) || CanUse(LONGSHOT) || (LogicColossusGS && CanUse(HOOKSHOT))) && CanGetNightTimeGS;}}),
```
```suggestion
                  LocationAccess(COLOSSUS_GS_HILL,          {[]{return IsAdult && AtNight && ((CanPlantBean(DESERT_COLOSSUS) && CanAdultAttack) || CanUse(LONGSHOT) || (LogicColossusGS && CanUse(HOOKSHOT))) && CanGetNightTimeGS;}}),
```

inspectredc commented on soh/soh/Enhancements/randomizer/3drando/location_access/locacc_ice_cavern.cpp:
```diff
@@ -24,7 +24,7 @@ void AreaTable_Init_IceCavern() {
                   //Exits
                   Entrance(ICE_CAVERN_ENTRYWAY, {[]{return true;}}),
-                  Entrance(ICE_CAVERN_MAIN,     {[]{return Here(ICE_CAVERN_BEGINNING, []{return IsAdult || HasExplosives || CanUse(DINS_FIRE);});}}),
+                  Entrance(ICE_CAVERN_MAIN,     {[]{return Here(ICE_CAVERN_BEGINNING, []{return (CanUse(MASTER_SWORD) || CanUse(BIGGORON_SWORD)) || HasExplosives || CanUse(DINS_FIRE);});}}),
```
```suggestion
                  Entrance(ICE_CAVERN_MAIN,     {[]{return Here(ICE_CAVERN_BEGINNING, []{return (CanUse(MASTER_SWORD) || CanUse(KOKIRI_SWORD) || CanUse(BIGGORON_SWORD)) || HasExplosives || CanUse(DINS_FIRE);});}}),
```

inspectredc commented on soh/soh/Enhancements/randomizer/3drando/location_access/locacc_ice_cavern.cpp:
```diff
@@ -53,24 +53,24 @@ void AreaTable_Init_IceCavern() {
                   //Exits
                   Entrance(ICE_CAVERN_ENTRYWAY,             {[]{return true;}}),
-                  Entrance(ICE_CAVERN_MQ_MAP_ROOM,          {[]{return IsAdult || CanUse(DINS_FIRE) || (HasExplosives && (CanUse(STICKS) || CanUse(SLINGSHOT) || KokiriSword));}}),
+                  Entrance(ICE_CAVERN_MQ_MAP_ROOM,          {[]{return CanUse(MASTER_SWORD) || CanUse(BIGGORON_SWORD) || CanUse(DINS_FIRE) || (HasExplosives && (CanUse(STICKS) || CanUse(SLINGSHOT) || CanJumpslash));}}),
```
```suggestion
                  Entrance(ICE_CAVERN_MQ_MAP_ROOM,          {[]{return CanUse(MASTER_SWORD) || CanUse(KOKIRI_SWORD) || CanUse(BIGGORON_SWORD) || CanUse(DINS_FIRE) || (HasExplosives && (CanUse(STICKS) || CanUse(SLINGSHOT) || CanUse(BOW)));}}),
```

CanJumpslash would only be checking for kokiri sword which ive moved to earlier anyway
I've put Bow here too but basing it off my memory right now

inspectredc commented on soh/soh/Enhancements/randomizer/3drando/location_access/locacc_lost_woods.cpp:
```diff
@@ -144,7 +144,7 @@ void AreaTable_Init_LostWoods() {
                   LocationAccess(LW_DEKU_SCRUB_NEAR_DEKU_THEATER_RIGHT, {[]{return IsChild && CanStunDeku;}}),
                   LocationAccess(LW_DEKU_SCRUB_NEAR_DEKU_THEATER_LEFT,  {[]{return IsChild && CanStunDeku;}}),
-                  LocationAccess(LW_GS_ABOVE_THEATER,                   {[]{return IsAdult && AtNight && (CanPlantBean(LW_BEYOND_MIDO) || (LogicLostWoodsGSBean && CanUse(HOOKSHOT) && (CanUse(LONGSHOT) || CanUse(BOW) || CanUse(SLINGSHOT) || HasBombchus || CanUse(DINS_FIRE)))) && CanGetNightTimeGS;}}),
+                  LocationAccess(LW_GS_ABOVE_THEATER,                   {[]{return IsAdult && CanAdultAttack && AtNight && (CanPlantBean(LW_BEYOND_MIDO) || (LogicLostWoodsGSBean && CanUse(HOOKSHOT) && (CanUse(LONGSHOT) || CanUse(BOW) || CanUse(SLINGSHOT) || HasBombchus || CanUse(DINS_FIRE)))) && CanGetNightTimeGS;}}),
```
```suggestion
                  LocationAccess(LW_GS_ABOVE_THEATER,                   {[]{return IsAdult && AtNight && ((CanPlantBean(LW_BEYOND_MIDO) && CanAdultAttack) || (LogicLostWoodsGSBean && CanUse(HOOKSHOT) && (CanUse(LONGSHOT) || CanUse(BOW) || CanUse(SLINGSHOT) || HasBombchus || CanUse(DINS_FIRE)))) && CanGetNightTimeGS;}}),
```

inspectredc commented on soh/soh/Enhancements/randomizer/3drando/location_access/locacc_lost_woods.cpp:
```diff
@@ -218,7 +218,7 @@ void AreaTable_Init_LostWoods() {
   areaTable[SFM_WOLFOS_GROTTO] = Area("SFM Wolfos Grotto", "SFM Wolfos Grotto", NONE, NO_DAY_NIGHT_CYCLE, {}, {
                   //Locations
-                  LocationAccess(SFM_WOLFOS_GROTTO_CHEST, {[]{return IsAdult || Slingshot || Sticks || KokiriSword || CanUse(DINS_FIRE) || CanUse(MEGATON_HAMMER) || CanUse(MASTER_SWORD) || CanUse(BIGGORON_SWORD);}}),
+                  LocationAccess(SFM_WOLFOS_GROTTO_CHEST, {[]{return CanUse(SLINGSHOT) || CanUse(STICKS) || CanUse(KOKIRI_SWORD) || CanUse(DINS_FIRE) || CanUse(MEGATON_HAMMER) || CanUse(MASTER_SWORD) || CanUse(BIGGORON_SWORD);}}),
```
```suggestion
                  LocationAccess(SFM_WOLFOS_GROTTO_CHEST, {[]{return CanJumpslash || CanUse(SLINGSHOT) || CanUse(BOW) || CanUse(DINS_FIRE) || CanUse(MEGATON_HAMMER);}}),
```

inspectredc commented on soh/soh/Enhancements/randomizer/3drando/location_access/locacc_lost_woods.cpp:
```diff
@@ -185,7 +185,7 @@ void AreaTable_Init_LostWoods() {
                   //Exits
                   Entrance(LW_BEYOND_MIDO,       {[]{return true;}}),
-                  Entrance(SACRED_FOREST_MEADOW, {[]{return IsAdult || Slingshot || Sticks || KokiriSword || CanUse(DINS_FIRE) || CanUse(MEGATON_HAMMER) || CanUse(MASTER_SWORD) || CanUse(BIGGORON_SWORD);}}),
+                  Entrance(SACRED_FOREST_MEADOW, {[]{return CanUse(SLINGSHOT) || CanUse(STICKS) || CanUse(KOKIRI_SWORD) || CanUse(DINS_FIRE) || CanUse(MEGATON_HAMMER) || CanUse(MASTER_SWORD) || CanUse(BIGGORON_SWORD);}}),
```
```suggestion
                  Entrance(SACRED_FOREST_MEADOW, {[]{return IsAdult || CanJumpslash || CanUse(SLINGSHOT) || CanUse(BOW) || CanUse(DINS_FIRE) || CanUse(MEGATON_HAMMER);}}),
```
same as with the wolfos grotto but in this case adult can get through itemless since there is no wolfos + gate blocking the path. and adult can get through the maze itemless also

inspectredc commented on soh/soh/Enhancements/randomizer/3drando/location_access/locacc_spirit_temple.cpp:
```diff
@@ -46,9 +46,9 @@ void AreaTable_Init_SpiritTemple() {
                   LocationAccess(SPIRIT_TEMPLE_CHILD_CLIMB_EAST_CHEST,  {[]{return HasProjectile(HasProjectileAge::Both) || ((SmallKeys(SPIRIT_TEMPLE, 3) || (SmallKeys(SPIRIT_TEMPLE, 2) && BombchusInLogic && ShuffleDungeonEntrances.Is(SHUFFLEDUNGEONS_OFF))) && CanUse(SILVER_GAUNTLETS) && HasProjectile(HasProjectileAge::Adult)) || (SmallKeys(SPIRIT_TEMPLE, 5) && IsChild && HasProjectile(HasProject ...
                   LocationAccess(SPIRIT_TEMPLE_GS_SUN_ON_FLOOR_ROOM,    {[]{return HasProjectile(HasProjectileAge::Both) || CanUse(DINS_FIRE) ||
-                                                                                      (CanTakeDamage && (Sticks || KokiriSword || HasProjectile(HasProjectileAge::Child))) ||
+                                                                                      (CanTakeDamage && ((CanJumpslash) || HasProjectile(HasProjectileAge::Child))) ||
```
```suggestion
                                                                                      (CanTakeDamage && (CanJumpslash || HasProjectile(HasProjectileAge::Child))) ||
```

inspectredc commented on soh/soh/Enhancements/randomizer/3drando/location_access/locacc_zoras_domain.cpp:
```diff
@@ -95,7 +95,7 @@ void AreaTable_Init_ZorasDomain() {
                   LocationAccess(ZD_KING_ZORA_THAWED,    {[]{return KingZoraThawed;}}),
                   LocationAccess(ZD_TRADE_PRESCRIPTION,  {[]{return KingZoraThawed && Prescription;}}),
-                  LocationAccess(ZD_GS_FROZEN_WATERFALL, {[]{return IsAdult && AtNight && (HookshotOrBoomerang || CanUse(SLINGSHOT) || Bow || MagicMeter || LogicDomainGS) && CanGetNightTimeGS;}}),
+                  LocationAccess(ZD_GS_FROZEN_WATERFALL, {[]{return IsAdult && AtNight && (HookshotOrBoomerang || CanUse(SLINGSHOT) || Bow || (MagicMeter && (MasterSword || BiggoronSword)) || LogicDomainGS) && CanGetNightTimeGS;}}),
```
```suggestion
                  LocationAccess(ZD_GS_FROZEN_WATERFALL, {[]{return IsAdult && AtNight && (HookshotOrBoomerang || CanUse(SLINGSHOT) || Bow || (MagicMeter && (CanUse(MASTER_SWORD) || CanUse(KOKIRI_SWORD) || CanUse(BIGGORON_SWORD))) || LogicDomainGS) && CanGetNightTimeGS;}}),
```

inspectredc reviewed (changes requested):
Few changes here, I realised when suggesting the hammer gold skull changes that it's definitely outside the scope of this PR to address those so feel free to not change them. 

Mostly just adding some kokiri sword as adult checks here and there. If not hopefully I've left enough comments or it's clear enough as is

stratomaster64 committed:
Apply (most) locacc review suggestions

Co-authored-by: inspectredc <78732756+inspectredc@users.noreply.github.com>

stratomaster64 committed:
Reorganized swordless check for interface to fit edge cases;
getting master sword no longer highlights box

stratomaster64 commented:
I posted this in Discord, but will also post here for visibility:
Master Sword will *not* be compatible with the "Ask to Equip" enhancement. This is because, currently, the enhancement checks for specific text IDs to inject the textbox in `z_message_PAL.c` as well as whether the item received is not a rando-specific item, in `z_player.c`. Although the latter can probably be an easy fix, the issue is more apparent with the first one.
Rando-specific items (like MS or Greg) have a special text id, `TEXT_RANDOMIZER_CUSTOM_ITEM`, which does not contain a special message but has to be fetched from a custom message table. By comparison, vanilla equipment have designated get item text IDs that are easily distinguishable to inject the "equip now" prompt.

jbodner09 commented:
@stratomaster64 it may be worth opening up an issue to track that since it's a reasonable request to have it work like the rest of the equipment does and we can at least point people to the issue so they know we're aware of it when they complain.

stratomaster64 committed:
Merge branch 'develop' into mister-sword

Commit:
Merge branch 'HarbourMasters:develop' into mister-sword

stratomaster64 committed:
Edge case for BGS but no bow

Commit:
Merge branch 'HarbourMasters:develop' into mister-sword

lilacLunatic committed:
Fix implicit declaration error for GI hooks (#9)

Commit:
Merge branch 'HarbourMasters:develop' into mister-sword

Commit:
Merge branch 'develop' into mister-sword

inspectredc commented on soh/soh/Enhancements/randomizer/3drando/logic.cpp:
```diff
@@ -561,8 +562,10 @@ namespace Logic {
-    CanStunDeku     = IsAdult || CanChildAttack || Nuts || HasShield;
-    CanCutShrubs    = IsAdult /*|| Sticks*/ || KokiriSword || Boomerang || HasExplosives || CanUse(MASTER_SWORD) || CanUse(MEGATON_HAMMER) || CanUse(BIGGORON_SWORD);
+    CanAdultAttack  = IsAdult && (CanUse(SLINGSHOT) || CanUse(BOOMERANG) || CanUse(STICKS) || CanUse(KOKIRI_SWORD) || HasExplosives || CanUse(DINS_FIRE) || MasterSword || Hammer || BiggoronSword || Hookshot);
+    CanAdultDamage  = IsAdult && (CanUse(SLINGSHOT) ||                      CanUse(STICKS) || CanUse(KOKIRI_SWORD) || HasExplosives || CanUse(DINS_FIRE) || MasterSword || Hammer || BiggoronSword);
```
I believe this should have a reference to bow?

inspectredc commented on soh/soh/Enhancements/randomizer/3drando/location_access/locacc_fire_temple.cpp:
```diff
@@ -203,7 +204,7 @@ void AreaTable_Init_FireTemple() {
   areaTable[FIRE_TEMPLE_SCARECROW_ROOM] = Area("Fire Temple Scarecrow Room", "Fire Temple", FIRE_TEMPLE, NO_DAY_NIGHT_CYCLE, {}, {
                   //Locations
-                  LocationAccess(FIRE_TEMPLE_GS_SCARECROW_CLIMB, {[]{return IsAdult || CanChildAttack;}}),
+                  LocationAccess(FIRE_TEMPLE_GS_SCARECROW_CLIMB, {[]{return CanAdultAttack || CanChildAttack;}}),
```
```suggestion
                  LocationAccess(FIRE_TEMPLE_GS_SCARECROW_CLIMB, {[]{return CanJumpslash || CanUse(SLINGSHOT) || CanUse(BOOMERANG) || HasExplosives || CanUse(BOW) || CanUse(HOOKSHOT) || CanUse(DINS_FIRE);}}),
```
Another scenario where you cannot get the skulltula with the hammer I believe

inspectredc reviewed (changes requested):
Just had a quick scan through some of the logic again, will do some more later when i can

inspectredc commented on soh/soh/Enhancements/randomizer/3drando/location_access/locacc_gerudo_valley.cpp:
```diff
@@ -104,13 +104,13 @@ void AreaTable_Init_GerudoValley() {
                   LocationAccess(GF_GERUDO_MEMBERSHIP_CARD,       {[]{return CanFinishGerudoFortress;}}),
                   LocationAccess(GF_GS_ARCHERY_RANGE,   {[]{return IsAdult && HookshotOrBoomerang && GerudoToken && AtNight && CanGetNightTimeGS;}}),
-                  LocationAccess(GF_GS_TOP_FLOOR,       {[]{return IsAdult && AtNight && (GerudoToken || CanUse(BOW) || CanUse(HOOKSHOT) || CanUse(HOVER_BOOTS) || LogicGerudoKitchen || LogicGFJump) && CanGetNightTimeGS;}}),
+                  LocationAccess(GF_GS_TOP_FLOOR,       {[]{return IsAdult && AtNight && CanAdultAttack && (GerudoToken || CanUse(BOW) || CanUse(HOOKSHOT) || CanUse(HOVER_BOOTS) || LogicGerudoKitchen || LogicGFJump) && CanGetNightTimeGS;}}),
```
```suggestion
                  LocationAccess(GF_GS_TOP_FLOOR,       {[]{return IsAdult && AtNight && (CanJumpslash || CanUse(SLINGSHOT) || CanUse(BOOMERANG) || HasExplosives || CanUse(BOW) || CanUse(HOOKSHOT) || CanUse(DINS_FIRE)) && (GerudoToken || CanUse(BOW) || CanUse(HOOKSHOT) || CanUse(HOVER_BOOTS) || LogicGerudoKitchen || LogicGFJump) && CanGetNightTimeGS;}}),
```

Thought about it some more and would like to avoid any incorrect logic which could lead to softlocking the seed

inspectredc commented on soh/soh/Enhancements/randomizer/3drando/location_access/locacc_gerudo_valley.cpp:
```diff
@@ -149,7 +149,7 @@ void AreaTable_Init_GerudoValley() {
                   //Locations
                   LocationAccess(WASTELAND_CHEST,            {[]{return HasFireSource;}}),
-                  LocationAccess(WASTELAND_BOMBCHU_SALESMAN, {[]{return AdultsWallet && (IsAdult || Sticks || KokiriSword);}}),
+                  LocationAccess(WASTELAND_BOMBCHU_SALESMAN, {[]{return AdultsWallet && CanJumpslash;}}),
```
```suggestion
                  LocationAccess(WASTELAND_BOMBCHU_SALESMAN, {[]{return AdultsWallet && (CanJumpslash || CanUse(HOVER_BOOTS)) ;}}),
```

inspectredc commented on soh/soh/Enhancements/randomizer/3drando/location_access/locacc_ice_cavern.cpp:
```diff
@@ -34,8 +34,8 @@ void AreaTable_Init_IceCavern() {
-                  LocationAccess(ICE_CAVERN_IRON_BOOTS_CHEST,        {[]{return BlueFire && (IsAdult || Slingshot || Sticks || KokiriSword || CanUse(DINS_FIRE));}}),
-                  LocationAccess(SHEIK_IN_ICE_CAVERN,                {[]{return BlueFire && IsAdult;}}),
+                  LocationAccess(ICE_CAVERN_IRON_BOOTS_CHEST,        {[]{return BlueFire && (CanJumpslash || CanUse(SLINGSHOT) || CanUse(DINS_FIRE));}}),
+                  LocationAccess(SHEIK_IN_ICE_CAVERN,                {[]{return BlueFire && (CanJumpslash || CanUse(SLINGSHOT) || CanUse(DINS_FIRE)) && IsAdult;}}),
```
```suggestion
                  LocationAccess(ICE_CAVERN_IRON_BOOTS_CHEST,        {[]{return BlueFire && (CanJumpslash || CanUse(SLINGSHOT) || CanUse(BOW) || CanUse(MEGATON_HAMMER) || CanUse(DINS_FIRE));}}),
                  LocationAccess(SHEIK_IN_ICE_CAVERN,                {[]{return BlueFire && (CanJumpslash || CanUse(SLINGSHOT) || CanUse(BOW) || CanUse(MEGATON_HAMMER) || CanUse(DINS_FIRE)) && IsAdult;}}),
```
Logic was trying to cover these scenarios with IsAdult before, I think only bow and hammer are the additional items to add here

inspectredc commented on soh/soh/Enhancements/randomizer/3drando/location_access/locacc_lost_woods.cpp:
```diff
@@ -34,8 +34,8 @@ void AreaTable_Init_LostWoods() {
-                  EventAccess(&DekuBabaSticks,           {[]{return DekuBabaSticks || ((IsAdult && ShuffleDungeonEntrances.Is(SHUFFLEDUNGEONS_OFF)) || KokiriSword || Boomerang);}}),
-                  EventAccess(&DekuBabaNuts,             {[]{return DekuBabaNuts   || ((IsAdult && ShuffleDungeonEntrances.Is(SHUFFLEDUNGEONS_OFF)) || KokiriSword || Slingshot || Sticks || HasExplosives || CanUse(DINS_FIRE));}}),
+                  EventAccess(&DekuBabaSticks,           {[]{return DekuBabaSticks || ((IsAdult && (MasterSword || BiggoronSword) && ShuffleDungeonEntrances.Is(SHUFFLEDUNGEONS_OFF)) || KokiriSword || Boomerang);}}),
+                  EventAccess(&DekuBabaNuts,             {[]{return DekuBabaNuts   || ((IsAdult && (MasterSword || BiggoronSword) && ShuffleDungeonEntrances.Is(SHUFFLEDUNGEONS_OFF)) || KokiriSword || Slingshot || Sticks || HasExplosives || CanUse(DINS_FIRE));}}),
```
```suggestion
                  EventAccess(&DekuBabaSticks,           {[]{return DekuBabaSticks || ((IsAdult && (CanUse(MASTER_SWORD) || CanUse(BIGGORON_SWORD) || CanUse(KOKIRI_SWORD)) && !ShuffleEntrances) || (IsChild && (CanUse(KOKIRI_SWORD) || CanUse(BIGGORON_SWORD) || CanUse(MASTER_SWORD) || CanUse(BOOMERANG))));}}),
                  EventAccess(&DekuBabaNuts,             {[]{return DekuBabaNuts   || ((IsAdult && (CanUse(MASTER_SWORD) || CanUse(BIGGORON_SWORD) || CanUse(KOKIRI_SWORD)) && !ShuffleEntrances) || (IsChild && (CanJumpslash || CanUse(SLINGSHOT) || HasExplosives || CanUse(DINS_FIRE))));}}),
```
this was already not 100% accurate before so just cleaning up

inspectredc commented on soh/soh/Enhancements/randomizer/3drando/location_access/locacc_spirit_temple.cpp:
```diff
@@ -57,7 +57,7 @@ void AreaTable_Init_SpiritTemple() {
                   //Locations
                   LocationAccess(SPIRIT_TEMPLE_COMPASS_CHEST,            {[]{return CanUse(HOOKSHOT) && CanPlay(ZeldasLullaby);}}),
-                  LocationAccess(SPIRIT_TEMPLE_EARLY_ADULT_RIGHT_CHEST,  {[]{return Bow || Hookshot || HasBombchus || (Bombs && LogicSpiritLowerAdultSwitch);}}),
+                  LocationAccess(SPIRIT_TEMPLE_EARLY_ADULT_RIGHT_CHEST,  {[]{return (Bow || Hookshot || HasBombchus || (Bombs && LogicSpiritLowerAdultSwitch)) && (HoverBoots || CanJumpslash);}}),
```
```suggestion
                  LocationAccess(SPIRIT_TEMPLE_EARLY_ADULT_RIGHT_CHEST,  {[]{return (CanUse(BOW) || CanUse(HOOKSHOT) || CanUse(SLINGSHOT) || CanUse(BOOMERANG) || HasBombchus || (Bombs && IsAdult && LogicSpiritLowerAdultSwitch)) && (CanUse(HOVER_BOOTS) || CanJumpslash);}}),
```

inspectredc commented on soh/soh/Enhancements/randomizer/3drando/location_access/locacc_water_temple.cpp:
```diff
@@ -122,7 +122,7 @@ void AreaTable_Init_WaterTemple() {
   areaTable[WATER_TEMPLE_SOUTH_LOWER] = Area("Water Temple South Lower", "Water Temple", WATER_TEMPLE, NO_DAY_NIGHT_CYCLE, {}, {
                   //Locations
-                  LocationAccess(WATER_TEMPLE_GS_BEHIND_GATE, {[]{return CanUse(HOOKSHOT) || (IsAdult && CanUse(HOVER_BOOTS));}}),
+                  LocationAccess(WATER_TEMPLE_GS_BEHIND_GATE, {[]{return CanUse(HOOKSHOT) || (IsAdult && (CanJumpslash || CanUseProjectile) && CanUse(HOVER_BOOTS));}}),
```
```suggestion
                  LocationAccess(WATER_TEMPLE_GS_BEHIND_GATE, {[]{return (CanJumpslash || CanUse(MEGATON_HAMMER)) && (CanUse(HOOKSHOT) || (IsAdult && CanUse(HOVER_BOOTS)));}}),
```
you cant get past the gate without a jumpslash item as far as i can tell (even explosives dont reach)

inspectredc commented on soh/soh/Enhancements/randomizer/3drando/location_access/locacc_zoras_domain.cpp:
```diff
@@ -95,7 +95,7 @@ void AreaTable_Init_ZorasDomain() {
                   LocationAccess(ZD_KING_ZORA_THAWED,    {[]{return KingZoraThawed;}}),
                   LocationAccess(ZD_TRADE_PRESCRIPTION,  {[]{return KingZoraThawed && Prescription;}}),
-                  LocationAccess(ZD_GS_FROZEN_WATERFALL, {[]{return IsAdult && AtNight && (HookshotOrBoomerang || CanUse(SLINGSHOT) || Bow || MagicMeter || LogicDomainGS) && CanGetNightTimeGS;}}),
+                  LocationAccess(ZD_GS_FROZEN_WATERFALL, {[]{return IsAdult && AtNight && (HookshotOrBoomerang || CanUse(SLINGSHOT) || Bow || (MagicMeter && (CanUse(MASTER_SWORD) || CanUse(KOKIRI_SWORD) || CanUse(BIGGORON_SWORD))) || LogicDomainGS) && CanGetNightTimeGS;}}),
```
```suggestion
                  LocationAccess(ZD_GS_FROZEN_WATERFALL, {[]{return IsAdult && AtNight && (HookshotOrBoomerang || CanUse(SLINGSHOT) || Bow || (MagicMeter && (CanUse(MASTER_SWORD) || CanUse(KOKIRI_SWORD) || CanUse(BIGGORON_SWORD))) || (LogicDomainGS && CanJumpslash)) && CanGetNightTimeGS;}}),
```
the trick requires a jumpslash

inspectredc reviewed (changes requested):
Finished going through the best I could. Mostly I have tried to highlight cases which could lead to checks incorrectly being 'in logic' when the player actually has no way of obtaining them glitchlessly

stratomaster64 committed:
Adjusted `CanAdultAttack/Damage`; applied logic suggestions

Commit:
Merge branch 'HarbourMasters:develop' into mister-sword

Commit:
Merge branch 'develop' into mister-sword

Commit:
Merge branch 'develop' into mister-sword

Commit:
Merge branch 'HarbourMasters:develop' into mister-sword

stratomaster64 committed:
Fixed build errors (hopefully)

Commit:
Merge branch 'develop' into mister-sword

Commit:
Merge branch 'develop' into mister-sword

stratomaster64 committed:
Cleanup merge

stratomaster64 committed:
get shit working again

Commit:
Merge branch 'develop' into mister-sword

stratomaster64 committed:
Tidied up remaining uses of DD flag as rando indicator

Archez committed:
make master sword invisible and fix ms flag (#10)

stratomaster64 committed:
Add text to sheik if go mode is obtained but barrier is still up

stratomaster64 committed:
overhaul swordless behavior in `func_80083108`

stratomaster64 committed:
reworked ToT MS Check to have an actual GI

Archez commented on soh/src/overlays/actors/ovl_Bg_Toki_Swd/z_bg_toki_swd.c:
```diff
@@ -122,7 +123,9 @@ void func_808BAF40(BgTokiSwd* this, PlayState* play) {
         if (Actor_HasParent(&this->actor, play)) {
             if (!LINK_IS_ADULT) {
-                Item_Give(play, ITEM_SWORD_MASTER);
+                 if (!Randomizer_GetSettingValue(RSK_SHUFFLE_MASTER_SWORD)) {
+                    Item_Give(play, ITEM_SWORD_MASTER);
+                 }
```
Vanilla code can enter this if block due to the `!IS_RANDO` above, and we need to avoid using `Randomizer_GetSettingValue` in a vanilla code flow. Should add another `!IS_RANDO` check here:

```suggestion
                if (!IS_RANDO || !Randomizer_GetSettingValue(RSK_SHUFFLE_MASTER_SWORD)) {
                    Item_Give(play, ITEM_SWORD_MASTER);
                }
```

Archez commented on soh/src/overlays/actors/ovl_Bg_Toki_Swd/z_bg_toki_swd.c:
```diff
@@ -6,6 +6,7 @@
 
 #include "z_bg_toki_swd.h"
 #include "objects/object_toki_objects/object_toki_objects.h"
+#include "soh/Enhancements/game-interactor/GameInteractor_Hooks.h"
```
Import is no longer needed

```suggestion
```

stratomaster64 committed:
suggestions

Commit:
Merge branch 'develop' into mister-sword

Pepe20129 commented on soh/soh/Enhancements/randomizer/randomizer.cpp:
```diff
@@ -3832,6 +3839,20 @@ void RandomizerSettingsWindow::DrawElement() {
 
+                //Shuffle Master Sword
+                //RANDOTODO: Disable when Start with Master Sword is active
+                // bool disableShuffleMasterSword = CvarGetInteger("gRandomizeStartingMasterSword", 0);
```
Looks like this ended up being unused

stratomaster64 replied:
May consider leaving this in for when the Starting Inventory section is fleshed out.

Pepe20129 reviewed (approved):
Did a few randos and everything was functional; the only things I'd say are that when turning into adult, link visually pulls out the master sword and then stashes it away even if you don't have it, if it's possible I think it'd be better if either the sword was invisible or the animation was changed to another one (standing for example) and that in the ganon fight something similar happens where link visually takes out the sword even if you don't have it before ganon knocks it away

Those are just minor nitpicks tho and I think this is perfectly fine to merge

Archez commented on soh/src/overlays/actors/ovl_player_actor/z_player.c:
```diff
@@ -14925,10 +14925,13 @@ void func_80852648(PlayState* play, Player* this, CsCmdActorAction* arg2) {
+            Inventory_ChangeEquipment(EQUIP_SWORD, 2);
+            gSaveContext.equips.buttonItems[0] = ITEM_SWORD_MASTER;
+            Inventory_DeleteEquipment(play, 0);
+        }
```
Looks like this code is responsible for making link swordless and removing the MS sword in the ganon fight during the cutscene. As it stands with this change, if you had the big gorron sword equipped you would resume control of link with the BGS sword still assigned to the B button.

To match the vanilla experience, we should probably unqeuip whatever sword link has and set the swordless flag. I'm not sure the scene check is necessary, I think this code is only executed for the ganon cutscene anyways. Something link this:

```suggestion
        this->leftHandDLists = gPlayerLeftHandOpenDLs;
        
        // If MS sword is shuffled and not in the players inventory, then we need to unequip the current sword
        // and set swordless flag to mimic Link having his weapon knocked out of his hand in the Ganon fight
        if (IS_RANDO && Randomizer_GetSettingValue(RSK_SHUFFLE_MASTER_SWORD) && !CHECK_OWNED_EQUIP(EQUIP_SWORD, 1)) {
            Inventory_ChangeEquipment(EQUIP_SWORD, 0);
            gSaveContext.equips.buttonItems[0] = ITEM_NONE;
            Flags_SetInfTable(INFTABLE_SWORDLESS);
            return;
        }
        
        Inventory_ChangeEquipment(EQUIP_SWORD, 2);
        gSaveContext.equips.buttonItems[0] = ITEM_SWORD_MASTER;
        Inventory_DeleteEquipment(play, 0);
```

Archez commented on soh/src/code/z_play.c:
```diff
@@ -264,6 +264,18 @@ void GivePlayerRandoRewardRequiem(PlayState* play, RandomizerCheck check) {
+    Player* player = GET_PLAYER(play);
+
+    if (gSaveContext.entranceIndex == 0x02CA && LINK_IS_ADULT && !Player_InBlockingCsMode(play, player) &&
+    player != NULL && Randomizer_GetSettingValue(RSK_SHUFFLE_MASTER_SWORD) && !Flags_GetRandomizerInf(RAND_INF_TOT_MASTER_SWORD)) {
```
`player != NULL` should be checked before accessing player in `Player_InBlockingCsMode()`

```suggestion
    if (gSaveContext.entranceIndex == 0x02CA && LINK_IS_ADULT && player != NULL &&
        !Player_InBlockingCsMode(play, player) && Randomizer_GetSettingValue(RSK_SHUFFLE_MASTER_SWORD) &&
        !Flags_GetRandomizerInf(RAND_INF_TOT_MASTER_SWORD)) {
```

Archez reviewed (commented):
Looking good so far, very close to the finish line! I have provided a few more suggestions/fixes below.

Other than that, the only thing that still has me uneasy is all the swordless management, especially knowing there is still some issues with it regarding Epona. At the moment, I don't really have an idea on how to address that, so may just need to think on that for a bit.

Malkierian commented on soh/soh/Enhancements/randomizer/3drando/location_access/locacc_ganons_castle.cpp:
```diff
@@ -86,8 +86,8 @@ void AreaTable_Init_GanonsCastle() {
+                  EventAccess(&SpiritTrialClear, {[]{return CanUse(LIGHT_ARROWS) && (CanUse(MIRROR_SHIELD) || SunlightArrows) && HasBombchus && ((LogicSpiritTrialHookshot && CanJumpslash) || CanUse(HOOKSHOT));}}),
                 }, {
                   //Locations
                   LocationAccess(GANONS_CASTLE_SPIRIT_TRIAL_CRYSTAL_SWITCH_CHEST, {[]{return LogicSpiritTrialHookshot || CanUse(HOOKSHOT);}}),
```
So unless you can hit this thing's crystal switch with the hookshot, you might need to include logic conditions for having a sword as adult. I think it assumes right now that being able to use hookshot also means you have a sword, since you'd be adult, and a spin attack will hit it through the wall.

inspectredc replied:
yeah the line would need to be
```LocationAccess(GANONS_CASTLE_SPIRIT_TRIAL_CRYSTAL_SWITCH_CHEST, {[]{return (LogicSpiritTrialHookshot || CanUse(HOOKSHOT)) && CanJumpslash;}}),```
I imagine

stratomaster64 committed:
Apply suggestions

Commit:
Merge branch 'HarbourMasters:develop' into mister-sword

Archez committed:
Better swordless handling with temp B (#11)

* better swordless handling with temp B

* prevent auto save in fishing pond

* prevent auto save during bombchu bowling

Archez commented:
Copying description from https://github.com/stratomaster64/Shipwright/pull/11 for visibility/history

---

This PR includes improvements for swordless state handling during minigames/epona by hijacking tempB with a "marker" value to allow us to restore the B button as empty without "disabling" the B button in the HUD. This should resolve all remaining issues with the HUD.
This also resolves child not having slingshot on B during child archery when swordless.

Additionally, an old swordless hack for child fishing can be removed in favor of the new handling, which should remove any possible cases that a kokiri sword is accidentally given to the player. This required preventing auto save in the fishing pond to avoid the fishing rod on B being saved to file.

A similar auto save restriction for bombchu bowling was needed. Shooting gallery did not have issues with auto save though.

The following scenarios have been tested:
* Riding Epona without a sword or bow
* Riding Epona with a sword but no bow
* Riding Epona without a sword but with a bow
* Pausing while on Epona
* Saving while on Epona
* Fishing as child or adult without a sword
* Pausing while fishing
* Saving while fishing
* Game over while fishing
* Playing shooting gallery as child or adult without a sword
* Game over during shooting gallery
* Playing bombchu bowling without a sword
* Game over during bomchu bowling

Behind the scenes, rather than storing `ITEM_NONE` which is the same as `BTN_DISABLED` (255), I've elected to store `ITEM_NONE_FE` (254) so that we have a marker to know that we were swordless before and it doesn't dim/disabled the B button like (255) does.

garrettjoecox reviewed (approved):
Bravo. I know this one sat in the oven for a while but it's turned out great and playtested quite a lot.

As discussed in the discord we need to swap the instances of `0x72` for `0xE0`, but assuming that is complete dropping an approval 👍

stratomaster64 committed:
enum fix

Commit:
Merge branch 'develop' into mister-sword

Archez reviewed (approved):
Beautiful! So excited to have this in. Thanks for working through the feedback.

If you get a chance, it would be nice to have something in the PR description listing roughly what was done for historical purposes, but not a blocker for merge 👍
